subsys/net/ip/dhcpv4.c

```diff
@@ -941,10 +945,20 @@ static void dhcpv4_handle_msg_nak(struct net_if *iface)
 	case NET_DHCPV4_DISABLED:
 	case NET_DHCPV4_INIT:
 	case NET_DHCPV4_SELECTING:
+	case NET_DHCPV4_REQUESTING:
+		if (memcmp(&request_server_addr, &iface->config.dhcpv4.server_id,
```

Comment:
I'm not sure if such an comparision is roboust enough. I see in RFC2132, that Server ID option, which updates the `config.dhcpv4.server_id` field is optional for DHCPNAK message. It could be the case, that the "wrong" server sends a NAK w/o this option, yet it'd still be processed. IMO it'd be better to rely on the source address present in the IP header.

And generally, probably not something to be fixed in this PR, but I find it actually awkward to process the DHCP options before it's decided whether to process or drop the packet.

Reply:
I think this works sufficiently. The key here is that in the selecting state, an offer message has to have the server id option set. So `server_id` will be the offer server which is what we want. If it is a nak, it will be empty which won't match or it will be the address of the naking server.

Reply:
Not sure what I was thinking, I see your point now. I fixed this. The src address of a response is used instead of the server_id field.